src/gpu_sysman.c

```diff
@@ -895,58 +983,87 @@ static bool gpu_mems_bw(gpu_device_t *gpu) {
     if (gpu->membw) {
       free(gpu->membw);
     }
-    gpu->membw = calloc(mem_count, sizeof(zes_mem_bandwidth_t));
+    gpu->membw = calloc(mem_count, sizeof(*gpu->membw));
     gpu->membw_count = mem_count;
     assert(gpu->membw);
   }
 
-  bool ok = false;
+  metric_family_t fam_ratio = {
+      .help = "Average memory bandwidth usage ratio (0-1) over query interval",
+      .name = METRIC_PREFIX "memory_bw_ratio",
+      .type = METRIC_TYPE_GAUGE,
+  };
+  metric_family_t fam_counter = {
+      .help = "Memory bandwidth usage total (in bytes)",
+      .name = METRIC_PREFIX "memory_bw_bytes_total",
+      .type = METRIC_TYPE_COUNTER,
+  };
+  metric_t metric = {0};
+
+  bool reported_ratio = false, reported_counter = false, ok = false;
   for (i = 0; i < mem_count; i++) {
-    char prefix[32];
-    if (!set_mem_prefix(mems[i], prefix, sizeof(prefix))) {
-      ERROR(PLUGIN_NAME ": failed to get memory module %d properties", i);
-      ok = false;
-      break;
-    }
     ze_result_t ret;
     zes_mem_bandwidth_t bw;
     if (ret = zesMemoryGetBandwidth(mems[i], &bw), ret != ZE_RESULT_SUCCESS) {
-      ERROR(PLUGIN_NAME
-            ": failed to get memory module %d (%s) bandwidth => 0x%x",
-            i, prefix, ret);
+      ERROR(PLUGIN_NAME ": failed to get memory module %d bandwidth => 0x%x", i,
+            ret);
       ok = false;
       break;
     }
+    if (!set_mem_labels(mems[i], &metric)) {
+      ERROR(PLUGIN_NAME ": failed to get memory module %d properties", i);
+      ok = false;
+      break;
+    }
+    if (config.output & OUTPUT_RAW) {
+      metric.value.counter = bw.writeCounter;
+      metric_label_set(&metric, "direction", "write");
+      metric_family_metric_append(&fam_counter, metric);
+
+      metric.value.counter = bw.readCounter;
+      metric_label_set(&metric, "direction", "read");
+      metric_family_metric_append(&fam_counter, metric);
+      reported_counter = true;
+    }
     zes_mem_bandwidth_t *old = &gpu->membw[i];
-    if (old->maxBandwidth) {
+    if (old->maxBandwidth && (config.output & OUTPUT_DERIVED)) {
       /* https://spec.oneapi.com/level-zero/latest/sysman/api.html#_CPPv419zes_mem_bandwidth_t
        */
       uint64_t writes = bw.writeCounter - old->writeCounter;
       uint64_t reads = bw.readCounter - old->readCounter;
       uint64_t timediff = bw.timestamp - old->timestamp;
-      double factor = 100.0 * 1.0e6 / (bw.maxBandwidth * timediff);
-      char vname[64];
+      double factor = 1.0e6 / (bw.maxBandwidth * timediff);
```

Comment:
can we rest assured that there is no possibility for a div-by-zero here?

Reply:
The added commit handles that by comparing timestamps, and using old->maxBandwidth.